contrib/pkg/clusterpool/clusterpool.go

```diff
@@ -0,0 +1,464 @@
+		},
+		Spec: hivev1.ClusterImageSetSpec{
+			ReleaseImage: o.ReleaseImage,
+		},
+	}
+	o.ClusterImageSet = imageSet.Name
+	return imageSet, nil
+}
+
+func (o *ClusterPoolOptions) build() ([]runtime.Object, error) {
+	// validate
+	if len(o.Name) == 0 {
+		return nil, fmt.Errorf("name is required")
+	}
+	if len(o.BaseDomain) == 0 {
+		return nil, fmt.Errorf("BaseDomain is required")
+	}
+	if len(o.ClusterImageSet) == 0 && len(o.ReleaseImage) == 0 {
+		return nil, fmt.Errorf("must set either image set or release image")
+	}
+
+	var allObjects []runtime.Object
+
+	allObjects = append(allObjects, o.generateClusterPool())
+
+	if o.createCloudSecret {
+		cloudCredsSecret := o.generateCredentialsSecret()
+		if cloudCredsSecret != nil {
+			allObjects = append(allObjects, cloudCredsSecret)
+		}
+	}
+
+	return allObjects, nil
+}
+
+func (o *ClusterPoolOptions) getPullSecretSecretName() string {
+	return fmt.Sprintf("%s-pull-secret", o.Name)
+}
+
+func (o *ClusterPoolOptions) addClusterPoolPlatform(cp *hivev1.ClusterPool) {
+	switch o.Cloud {
+	case cloudAWS:
+		cp.Spec.Platform = hivev1.Platform{
+			AWS: &hivev1aws.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region: o.Region,
+			},
+		}
+	case cloudAzure:
+		cp.Spec.Platform = hivev1.Platform{
+			Azure: &hivev1azure.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region:                      o.Region,
+				BaseDomainResourceGroupName: o.AzureBaseDomainResourceGroupName,
+			},
+		}
+	case cloudGCP:
+		cp.Spec.Platform = hivev1.Platform{
+			GCP: &hivev1gcp.Platform{
+				CredentialsSecretRef: corev1.LocalObjectReference{
+					Name: o.credsSecretName(),
+				},
+				Region: o.Region,
+			},
+		}
+	}
+}
+
+func (o *ClusterPoolOptions) credsSecretName() string {
+	if o.createCloudSecret {
+		switch o.Cloud {
+		case cloudAWS:
+			return fmt.Sprintf("%s-aws-creds", o.Name)
+		case cloudAzure:
+			return fmt.Sprintf("%s-azure-creds", o.Name)
+		case cloudGCP:
+			return fmt.Sprintf("%s-gcp-creds", o.Name)
+		}
+		return o.CloudSecretName
+	}
+	return o.CloudSecretName
+}
+
+func (o *ClusterPoolOptions) generateClusterPool() *hivev1.ClusterPool {
+	cp := &hivev1.ClusterPool{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterPool",
+			APIVersion: hivev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        o.Name,
+			Annotations: map[string]string{},
+		},
+		Spec: hivev1.ClusterPoolSpec{
+			BaseDomain: o.BaseDomain,
+			Size:       o.Size,
+		},
+	}
+	if o.PullSecret != "" || o.PullSecretFile != "" {
+		cp.Spec.PullSecretRef = &corev1.LocalObjectReference{Name: o.getPullSecretSecretName()}
+	}
+	if o.ClusterImageSet != "" {
+		cp.Spec.ImageSetRef = hivev1.ClusterImageSetReference{Name: o.ClusterImageSet}
+	}
+	o.addClusterPoolPlatform(cp)
+
+	return cp
+}
+
+func (o *ClusterPoolOptions) generateCredentialsSecret() *corev1.Secret {
```

Comment:
I would like to see the `CloudBuilder` from `pkg/clusterresource` used to generate the creds secret rather than duplicating the code from there. Export the existing `generateCredentialsSecret` function.